prov/gni/src/gnix_ep.c

```diff
@@ -1874,6 +1917,106 @@ ssize_t gnix_cancel(fid_t fid, void *context)
 	return ret;
 }
 
+static int
+__gnix_ep_ops_get_val(struct fid *fid, dom_ops_val_t t, void *val)
+{
+	struct gnix_fid_ep *ep;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	assert(val);
+
+	if (fid->fclass != FI_CLASS_EP) {
+		GNIX_WARN(FI_LOG_DOMAIN, "Invalid ep\n");
+		return -FI_EINVAL;
+	}
+
+	ep = container_of(fid, struct gnix_fid_ep, ep_fid.fid);
+
+	switch (t) {
+	case GNI_HASH_TAG_IMPL:
+		*(uint32_t *)val = (ep->use_tag_hlist) ? 1 : 0;
+		break;
+
+	default:
+		GNIX_WARN(FI_LOG_DOMAIN, ("Invalid dom_ops_val\n"));
+		return -FI_EINVAL;
+	}
+
+	return FI_SUCCESS;
+}
+
+static int
+__gnix_ep_ops_set_val(struct fid *fid, dom_ops_val_t t, void *val)
+{
+	struct gnix_fid_ep *ep;
+	int v;
+	int ret;
+	int tag_type;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	assert(val);
+
+	if (fid->fclass != FI_CLASS_EP) {
+		GNIX_WARN(FI_LOG_DOMAIN, "Invalid ep\n");
+		return -FI_EINVAL;
+	}
+
+	ep = container_of(fid, struct gnix_fid_ep, ep_fid.fid);
+	switch (t) {
+	case GNI_HASH_TAG_IMPL:
+		if (ep->tx_enabled || ep->rx_enabled) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+					"EP enabled, cannot modify tag matcher\n");
+			return -FI_EINVAL;
+		}
+
+		v = *(uint32_t *) val;
+		if ((v && !(ep->use_tag_hlist)) ||
+				(!v && (ep->use_tag_hlist))) {
+			ret = __destruct_tag_storages(ep);
```

Comment:
The tagged storages are already allocated at this point. If the storages aren't allocated, the default state of the tag storage is 'destroyed', which means it would just return FI_EINVAL.